### PR TITLE
Module 7

### PR DESCRIPTION
**Pull request by reviakin**

### CODE REVIEW

File: src/app/repositories/user.ts [hunk 1]
```diff
@@ -10,10 +10,19 @@ class UserRepository implements IUserRepository {
     this.#userModel = model;
```
Comment (Kuznietsov) [inline src/app/repositories/user.ts]: If you are using TS, it is better to use its syntax for access modifiers. https://www.typescriptlang.org/docs/handbook/2/classes.html#private

File: src/app/routers/user.ts [hunk 2]
```diff
@@ -19,19 +21,26 @@ const BASE_ROUTE_PATH = "/users";
...
+    checkToken,
+    validateParamsForSuggestions,
+    userController.getSuggestions
+  )
   // get a user
-  .get("/:id", validateIdInParams, userController.getOneById)
+  .get("/:id", checkToken, validateIdInParams, userController.getOneById)
   // create a user
-  .post("/", validateCreateUserBody, userController.createOne)
+  .post("/", checkToken, validateCreateUserBody, userController.createOne)
```
Comment (Kuznietsov) [inline src/app/routers/user.ts]: Is it really necessary to check the token on user creation? It seems to me, this is a mistake. WDYT?

File: src/app/middlewares/checkToken.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  const token = getTokenFromRequestHeader(req);
+
+  if (!token) {
+    throw new Authorization("token is required");
+  }
+
+  try {
+    verify(token, JWT_SECRET);
+  } catch (error) {
+    throw new Forbidden("invalid token");
```
Comment (Kuznietsov) [inline src/app/middlewares/checkToken.ts]: At this place, you should process all kinds of errors, which can happen, send them to the client with the right HTTP status code. 

Possible errors: https://github.com/auth0/node-jsonwebtoken#errors--codes

File: src/utils/makeToken.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { sign } from "jsonwebtoken";
+import { JWT_SECRET } from "../config";
+
+function makeToken(payload: string | object | Buffer) {
```
Comment (Kuznietsov) [inline src/utils/makeToken.ts]: This function should be a part of the service for working with JWT. The refresh token can be added there, and also some other logic.

File: src/app/services/user.ts [hunk 5]
```diff
@@ -8,6 +10,21 @@ class UserService implements IUserService {
...
   }
 
+  login = async (username: string, password: string) => {
+    const user = await this.#userRepositories.getUserByLoginAndPassword(
+      username,
+      password
+    );
+
+    if (user) {
+      const token = makeToken(user);
```
Comment (Kuznietsov) [inline src/app/services/user.ts]: That is a very bad idea to save the user's password to the JWT token. At this point, you should pick only necessary fields to be saved to the token.

**Review by Kuznietsov — CHANGES_REQUESTED** [review]
Mostly, code LGTM. Could you, please, perform the requested changes?

File: src/app/repositories/user.ts (hunk 1 above)
Comment (reviakin) [reply src/app/repositories/user.ts]: fixed

File: src/app/middlewares/checkToken.ts (hunk 3 above)
Comment (reviakin) [reply src/app/middlewares/checkToken.ts]: fixed

File: src/utils/makeToken.ts (hunk 4 above)
Comment (reviakin) [reply src/utils/makeToken.ts]: the functionality has been moved to tokenService

**Review by Kuznietsov — APPROVED** [review]

**Pull request merged**